internal/controller/rbac/provider/roles/requests.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package roles
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Error strings.
+const (
+	errGetClusterRole = "cannot get ClusterRole"
+)
+
+const (
+	wildcard = "*"
+
+	pathPrefixURL      = "url"
+	pathPrefixResource = "resource"
+)
+
+// A Path represents a rule within the rule tree.
+type Path []string
+
+// A Node within the rule tree. Each node indicates whether the path should be
+// allowed. In practice only leaves (i.e. Verbs) should have this set.
+type Node struct {
+	allowed  bool
+	children map[string]*Node
+}
+
+// NewNode initialises and returns a Node.
+func NewNode() *Node {
+	return &Node{children: map[string]*Node{}}
+}
+
+// Allow the supplied path.
+func (n *Node) Allow(p Path) {
+	if len(p) == 0 {
+		n.allowed = true
+		return
+	}
+
+	k := p[0]
+	if _, ok := n.children[k]; !ok {
+		n.children[k] = NewNode()
+	}
+	n.children[k].Allow(p[1:])
+}
+
+// Allowed returns true if the supplied path is allowed.
+func (n *Node) Allowed(p Path) bool {
+	if len(p) == 0 {
+		return false
+	}
+
+	// TODO(negz): A NonResourceURL can't be a wildcard, but can have a final
+	// segment that is a wildcard; e.g. /apis/*. We don't currently support
+	// this.
+	for _, k := range []string{p[0], wildcard} {
+		if c, ok := n.children[k]; ok {
+			if c.allowed {
+				return true
+			}
+			if c.Allowed(p[1:]) {
+				return true
+			}
+		}
+	}
+	return false
+}
```

Comment:
Credit to @kasey for sketching out this tree implementation - does this look sound to you?

Reply:
Yeah I was on the fence about that too, but was too lazy to break this into a public method wrapping a private method for recursion. Arguably this whole implementation should be private.